### PR TITLE
Additions for app testing with Velocity

### PR DESCRIPTION
**Pull request by Sanjo**
Hi,

the purpose for this pull-request at this stage is to __get feedback__ from some of the Meteor core developers on the changes we have made for Velocity to enable app testing.

### Why

For velocity:core@0.6.0 we have created a fork of Meteor core in order to implement app testing in a more solid manner, in a similar way as how test-packages is implemented. This made Velocity more stable and reliable.

To make this possible, multiple changes to Meteor core were necessary. I will go through them one by one now.

### Changes

1. Separation of the projectDir and projectMeteorDir (The .meteor folder).
2. Option to include tests into the app for integration tests (--include-tests)
3. Command for testing the app, similar to `meteor test-packages` (`meteor run --test-app`)

#### 1. Separation of the projectDir and projectMeteorDir (The .meteor folder)

This allows to start the app multiple times from the same project directory. But each app instance has a separate .meteor folder. This change is required to enable the next two features.

#### 2. Option to include tests into the app for integration tests (--include-tests)

This allows us to include test files into the app. This way we can leverage the same Meteor build process for test files. We get the source code handlers and automatic restarts on changes.

#### 3. Command for testing the app

This command finally allows us to start the app in testing mode. We call this mirrors in Velocity. It works similar to `meteor test-packages` but it is for testing the whole app.

### How we use this features currently

In velocity:core 0.6.x we start mirrors with these changes ([see code](https://github.com/meteor-velocity/velocity/blob/v0.6.3/src/mirrors/Mirror.js#L241-L250)). We have released our own Meteor release and use it with --release.

This allowed us to experiment with these additions and we can make changes quickly without going through the Meteor release cycle.

The downside is that the user has to wait for the release to be downloaded the first time.

### Next steps

1. Get feedback from you
2. Discuss the roadmap for testing with Meteor

### CODE REVIEW

**Comment by stubailo** [comment]
This looks great, and it would be awesome if it made testing more stable by getting rid of the brittle mirroring code. I've taken a surface-level look and it looks pretty good to me so far. One question:

1. Why add a new option - `--test-app` instead of using the existing `test` option?

Some more questions inline.

File: tools/package-source.js [hunk 1]
```diff
@@ -1223,20 +1227,93 @@ _.extend(PackageSource.prototype, {
...
-                    /^cordova-build-override\/$/,
-                    otherUnibuildRegExp].concat(sourceExclude)
+          exclude: excludedTopLevelSubdirectories
         });
         checkForInfiniteRecursion('');
 
+        // Inputs that should return true:
+        // tests/...
+        // .../tests/...
+        var isTestsSubdirectory = function (dir) {
```
Comment (stubailo) [inline tools/package-source.js]: Can we factor these out into a file that has to do with testing? I'd rather not make this function even longer than it is now - 500 lines is really pushing it.

**Comment by stubailo** [comment]
Another question - you have added an option `projectMeteorDir` - do you ever use it to change something outside of `.meteor/local`? Could it be less powerful and be `projectLocalDir`, so that you don't end up copying the versions file, packages file, etc?

File: tools/package-source.js (hunk 1 above)
Comment (Sanjo) [inline tools/package-source.js]: Yeah.

**Comment by Sanjo** [comment]
> Why add a new option - `--test-app` instead of using the existing `test` option?

Because it does different things. `--test-app` is probably not a good name. What is does is basically starting the mirror for a specific testing framework. It's something that the user wouldn't use directly. For the user the test command for the CLI is still `meteor --test --release velocity:METEOR@1.1.0.2_3`.

> Another question - you have added an option projectMeteorDir - do you ever use it to change something outside of .meteor/local? Could it be less powerful and be projectLocalDir, so that you don't end up copying the versions file, packages file, etc?

That's a good point. No, we don't change anything outside of `.meteor/local`.

**Comment by stubailo** [comment]
So this is something that a testing framework would call internally? Then maybe a more "technical" name like `--velocity-mirror` or `--testing-mirror` or something?

**Comment by rissem** [comment]
I like `--testing-mirror`

**Comment by samhatoum** [comment]
+1 for --testing-mirror

**Comment by queso** [comment]
Is this PR dead? Looks like it has been maintained since the initial PR and should be good for 1.2?

**Comment by Sanjo** [comment]
Yes it works with Meteor 1.2.1. It's the branch that we use to run the Velocity mirrors.

The performance for app mirrors with larger apps is suboptimal because multiple running meteor build tools need quiet a lot of resources right now.

**Pull request closed without merging**